package protos;

// PROTOS FOR PACKAGING STRUCTS TO BE SENT
message vector {
	required float x = 1;
	required float y = 2;
	required float z = 3;
}

message RenderedObj {
	required uint32 id = 1;
	required uint32 mass = 2;
	required uint32 type = 3;
	required float size = 4;
	
	required float cur_tolerance = 5;
	required float max_tolerance = 6;
	required uint32 is_destroyed = 7;
	
	required vector pos = 8;
	required vector vel = 9;
	required float accel = 10;

	required vector roll_vector = 11;	
	required vector pitch_vector = 12;
	required vector yaw_vector = 13;
	required float pitch = 14;
	required float roll = 15;
	required float yaw = 16;
	
	optional uint32 weapon_index = 17;
}

message Event {
	required uint32 id = 1;
	required uint32 event_type = 2;
}

message ControlInput {
	required uint32 action = 1;
	optional float tilt = 2;  // if action == pitch/roll tilt, specify tilt
}



//PROTOS FOR CREATING NETWORK PACKETS
message GeneralPacket {
	required uint32 packetType = 1;  // contains id of packet type
	optional bytes packet = 2;  // serialized protobuf message that corresponds to packetType
	optional string packetId = 3;  // packet id
}

message ShipInitPacket {
	required RenderedObj ship = 1;
}

message ObjsAndEventsPacket {
	repeated RenderedObj obj = 1;
	repeated Event event = 2;
}

message ObjsAndEventsReqPacket {
	required uint32 req = 1;
}

message EventAckPacket {
	required uint32 ack = 1;
}

message ControlInputPacket {
	required ControlInput control = 1;
}
